### PR TITLE
Summary buckets and window operator parsing algorithm

### PR DESCRIPTION
**Pull request by ThomasQuesadilla**
Includes the definition for summary window operator and summary bucket structure.

***
<a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/pull/1111.html" title="Last updated on Dec 1, 2023, 9:36 PM UTC (7406c72)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/1111/b0aa8a3...7406c72.html" title="Last updated on Dec 1, 2023, 9:36 PM UTC (7406c72)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -677,6 +709,10 @@ A <dfn>trigger spec</dfn> is a [=struct=] with the following items:
 <dl dfn-for="trigger spec">
 : <dfn>event-level report windows</dfn>
 :: A [=report window list=].
+: A <dfn>summary window operator</dfn>
+::
```
Comment (apasel422) [inline index.bs:713]:
```suggestion
: <dfn>summary window operator</dfn>
:: A [=summary window operator=].
```

File: index.bs [hunk 2]
```diff
@@ -1866,6 +1902,28 @@ To <dfn>parse report windows</dfn> given a |value|, a
     1. Set |startDuration| to |endDuration|.
 1. Return |windows|.
 
+To <dfn>parse summary buckets</dfn> given a [=map=] |map|:
+1. Let |values| be the trivial mapping «[ 1, 2, ..., MAX_UINT32 ]».
```
Comment (apasel422) [inline index.bs:1906]: Can we use an [inclusive range](https://infra.spec.whatwg.org/#the-range) for this instead?

File: index.bs [hunk 3]
```diff
@@ -1866,6 +1902,28 @@ To <dfn>parse report windows</dfn> given a |value|, a
...
+    1. Let |summaryBucket| be a new [=summary bucket=] whose items are
+
+        : [=summary bucket/start=]
+        :: |prev|
+        : [=summary bucket/end=]
+        :: |item| - 1
+
+    1. [=list/Append=] |summaryBucket| to |summaryBuckets|.
+    1. Set |prev| to |item|.
+1. return |summaryBuckets|.
```
Comment (apasel422) [inline index.bs:1925]:
```suggestion
1. Return |summaryBuckets|.
```

File: index.bs [hunk 4]
```diff
@@ -1898,6 +1956,20 @@ To <dfn>parse trigger specs</dfn> given a [=map=] |map|, a [=moment=]
         1. If |reportWindows| is an error, return it.
         1. Set |spec|'s [=trigger spec/event-level report windows=] to
             |reportWindows|.
+    1. Let |summaryWindowOperator| be "<code>[=summary window operator/count=]</code>".
+    1. If |map|["`summary_window_operator`"] [=map/exists=]:
+        1. If |map|["`summary_window_operator`"] is not a [=string=], return an
+            error.
+        1. If |value|["`summary_window_operator`"] is not a
```
Comment (apasel422) [inline index.bs:1963]:
```suggestion
    1. If |item|["`summary_window_operator`"] [=map/exists=]:
        1. If |item|["`summary_window_operator`"] is not a [=string=], return an
            error.
        1. If |item|["`summary_window_operator`"] is not a 
```

File: index.bs [hunk 5]
```diff
@@ -1898,6 +1956,20 @@ To <dfn>parse trigger specs</dfn> given a [=map=] |map|, a [=moment=]
         1. If |reportWindows| is an error, return it.
         1. Set |spec|'s [=trigger spec/event-level report windows=] to
             |reportWindows|.
+    1. Let |summaryWindowOperator| be "<code>[=summary window operator/count=]</code>".
```
Comment (apasel422) [inline index.bs:1959]: Can we initialize the corresponding field in `|spec|` to this instead of introducing a local variable?

**Review by apasel422 — CHANGES_REQUESTED** [review]

File: index.bs [hunk 6]
```diff
@@ -1956,15 +1957,15 @@ To <dfn>parse trigger specs</dfn> given a [=map=] |map|, a [=moment=]
         1. If |reportWindows| is an error, return it.
         1. Set |spec|'s [=trigger spec/event-level report windows=] to
             |reportWindows|.
-    1. Let |summaryWindowOperator| be "<code>[=summary window operator/count=]</code>".
+    1. Set |spec|'s [=trigger spec/summary window operator=] to
```
Comment (apasel422) [inline index.bs:1960]: Sorry, I meant that we could do this as part of the `Let |spec| be a new [=trigger spec=]...` step above.

**Review by apasel422 — CHANGES_REQUESTED** [review]

File: index.bs (hunk 6 above)
Comment (ThomasQuesadilla) [reply index.bs:1960]: I'm assuming I should add the default values for these fields to "if map['trigger_specs'] does not exist" as well right?

**Review by apasel422 — CHANGES_REQUESTED** [review]

File: index.bs (hunk 6 above)
Comment (ThomasQuesadilla) [reply index.bs]: I removed the fields as well as the parsing implementation within the trigger spec algorithm. Please lmk if you meant something else, thanks!

File: index.bs [hunk 7]
```diff
@@ -1915,8 +1971,6 @@ To <dfn>parse trigger specs</dfn> given a [=map=] |map|, a [=moment=]
         1. Set |i| to |i| + 1.
 1. Return |specs|.
 
-Issue: Parse "`summary_window_operator`" and "`summary_buckets`" fields.
```
Comment (apasel422) [inline index.bs:1918]: Let's keep this issue but rephrase it as something like `Issue: Invoke [=parse summary buckets=] from this algorithm.`

File: index.bs [hunk 8]
```diff
@@ -655,6 +655,38 @@ Note: The [=report window list/total window=] is conceptually a union of
 [=report windows=], because there are no gaps in time between any of the
 [=report windows|windows=].
 
+<h3 id="summary-windows-operator-header">Summary window operator</h3>
+
+A <dfn>summary window operator</dfn> summarizes the triggers within a
```
Comment (apasel422) [inline index.bs:660]: Please also add the algorithm for parsing the `summary_window_operator` field.

**Review by apasel422 — CHANGES_REQUESTED** [review]

**Review by apasel422 — CHANGES_REQUESTED** [review]
Please update with respect to https://github.com/WICG/attribution-reporting-api/pull/1116.

**Review by apasel422 — CHANGES_REQUESTED** [review]

**Review by apasel422 — APPROVED** [review]

**Pull request merged**